language/move-vm/runtime/src/runtime.rs

```diff
@@ -49,19 +49,21 @@ impl VMRuntime {
         })
     }
 
+    pub(crate) fn mark_loader_cache_as_invalid(&self) {
+        self.loader.mark_as_invalid()
+    }
+
     pub fn new_session<'r, S: MoveResolver>(&self, remote: &'r S) -> Session<'r, '_, S> {
-        Session {
-            runtime: self,
-            data_cache: TransactionDataCache::new(remote, &self.loader),
-            native_extensions: NativeContextExtensions::default(),
-        }
+        self.new_session_with_extensions(remote, NativeContextExtensions::default())
     }
 
     pub fn new_session_with_extensions<'r, S: MoveResolver>(
         &self,
         remote: &'r S,
         native_extensions: NativeContextExtensions<'r>,
     ) -> Session<'r, '_, S> {
+        // If the loader cache is invalidated, flush it for the new session
+        self.loader.flush_if_invalidated();
```

Comment:
Should we do this automagically like this? 

I thought we would just poison it, block all VM session actions, and require the adapter to manually flush the cache

Reply:
I prefer to do this automatically. It's a bug in the VM and we should fix it on VM side. The real fix (hopefully coming soon) is rewrite the loader to be actually updatable.

Reply:
My concern is it could cause other silent failures. Lot's of endpoints (`load_script`, `load_function`, `load_type`, `get_type_layout`, etc) have pointers back into the cache. 
Invalidating the cache without warning the adapter I think is going to cause a lot of accidental errors here, unless I'm missing something.

Reply:
Ok, the assumption here was that no one has more than one session open at a time, but you are right that we can't enforce this. Removed the automatic flushing now and exposed another public function so the adapter can do the flushing.